### ISSUE TITLE
step-10 Q3 mapping issue

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#5039]
I am playing with step-10 and I noticed the following problem:
For Q3 the output according to the documentation should be 

![](https://www.dealii.org/images/steps/developer/step-10.ball_mapping_q3_ref0.png)

But this is what I get:

![step-10q3weird](https://user-images.githubusercontent.com/1531285/30135395-89bb5d36-9328-11e7-8079-a251dd0ef757.png)

Does anybody have an explanation for this?

Note that errors seem to be okay and Q2 looks reasonable.

**Comment by tjhei** [comment]
update: with ``DataOut`` and ``build_patches(large number)`` the result is correct. So this might be a bug in the gnuplot function for curved boundaries.

**Comment by bangerth** [comment]
Hm, that seems unlikely. gnuplot really only knows how to plot line segments. How did you create the second picture?

**Comment by tjhei** [comment]
> Hm, that seems unlikely. gnuplot really only knows how to plot line segments. How did you create the second picture?

per the instructions of step-10 in gnuplot: http://www.dealii.org/developer/doxygen/deal.II/step_10.html#Results

**Comment by tjhei** [comment]
with lines and points:
![image](https://user-images.githubusercontent.com/1531285/30136072-4c8e3a8e-932b-11e7-9495-0d3cd59dd5e0.png)

**Comment by bangerth** [comment]
This is the file that's generated by `GridOut::write_gnuplot`, right?

**Comment by tjhei** [comment]
Yes.

**Comment by kronbichler** [comment]
I have not looked at the actual numbers or the implementation yet. But the pattern in the points shows that points are more densely represented towards the corners, which appears to be a problem in the non-equidistant support points of `MappingQ` starting from Q3 and an assumption of equidistant points either in `write_gnuplot` or somewhere later (unlikely) that is somehow incompatible.

**Comment by kronbichler** [comment]
I had a quick look into the code of `write_gnuplot`
https://github.com/dealii/dealii/blob/master/source/grid/grid_out.cc#L3171
and we indeed use equidistant points there. However, we go through `transform_unit_to_real_cell` here
https://github.com/dealii/dealii/blob/master/source/grid/grid_out.cc#L3231
so I don't see how that could fail.

**Comment by kronbichler** [comment]
It would be interesting to see if the problem is in `transform_unit_to_real_cell` or somewhere else. I do not have the time to look into that now, but maybe later today or tomorrow.

**Comment by kronbichler** [comment]
@tjhei I had a closer look at this case and it is completely expected. The problem is `SphericalManifold`. Use `PolarManifold` and you will get a nice mesh - up to the point a cubic polynomial on four points can represent a circle, of course.

`SphericalManifold` has an error up to 4% from the actual radius, whereas `PolarManifold` offers an error of 0.3%, both on the same cubic representation. Of course, `SphericalManifold` gets better with a `MappingQ<2>(5)` (error is up to 0.5%) or a finer mesh, but if you compare it with `PolarManifold` on Q5, it is again dramatic (error less than 0.0015%). 

tl;dr We should replace `SphericalManifold` by `PolarManifold` in this example. `SphericalManifold` is OK on the sphere in 3D, but we have MUCH better options on the circle.

**Comment by tjhei** [comment]
Ah, thanks!
The issue is that the interpolation of this 90 degree chunk is not exact at the corners.

I just checked it, refining the mesh once or switching to the ``PolarManifold`` fixes the issue. Note that we say in the ``SphericalManifold`` documentation that "This Manifold is identical to PolarManifold in dimension two", which is not quite correct. ;-)

> tl;dr We should replace SphericalManifold by PolarManifold in this example. SphericalManifold is OK on the sphere in 3D, but we have MUCH better options on the circle.

This is sadly a little more complicated and would need to be explained in the description.

**Comment by kronbichler** [comment]
This was due to rewriting `SphericalManifold` last year in d1e3ffc through #2917. I think we should fix the comments in that file then, and clearly recommend to use `PolarManifold` in 2D if possible.

**Comment by tjhei** [comment]
> This was due to rewriting SphericalManifold last year in d1e3ffc through #2917. I think we should fix the comments in that file then, and clearly recommend to use PolarManifold in 2D if possible.

Makes sense. This is an annoying situation though, because it makes it harder to write dimension independent code. Considering how difficult it is to get the manifold for a sphere correct (see the step-6 discussion), I would say we should provide an easy way in the library to get a good hyper_sphere() setup. Maybe something like hyper_sphere_with_manifold() that creates/attaches the correct manifolds.

**Comment by kronbichler** [comment]
One thing we could do is to create a specialization `SphericalManifold<2>` that inherits from `PolarManifold`. This completely disables the abilities of `SphericalManifold` but I have fallen into the trap of `SphericalManifold<2>` when I wanted to have `PolarManifold` more than once.

**Comment by tjhei** [comment]
I don't like the fact that you get different behavior depending on the dimension used. This makes it harder to find bugs. As an example: if I do the plot like above in 3d, I would also get this ugly behavior around the corners, but I would not see the problem in 2d anymore.

@luca-heltai what do you think?

**Comment by luca-heltai** [comment]
During the last two years, I came to the conclusion that elementary geometry is one of the most difficult topic ever... 

I don't know what to think. The fact that `SphericalManifold` does not work so well in 2D is a good indication that we should change it. I skimmed through the documentation and through the code, and I don't recognize anymore what is happening there, so this will require more time. 

I have the impression that there is a bug in `SphericalManifold` anyway. What the documentation says *should* be true.

We compute the angle between two points, then split the angle according to weights, and construct the intermediate point. In my elementary geometry knowledge, splitting the angle and splitting the arch are the same operation. Polar manifold manipulates the angle directly as a coordinate, but this is essentially the same. Why it is different in practice, I have no idea, and I tend to believe this to be a bug.

**Comment by bangerth** [comment]
Has this been solved by the recent rewrite of the `SphericalManifold` function? @kronbichler @masterleinad ?

**Comment by drwells** [comment]
@bangerth my understanding is that the changes to support point calculation in #5386 fix this.

**Comment by masterleinad** [comment]
We are using `PolarManifold` in `SphericalManifold` for `spacedim==2` so I take the freedom to close.

**Comment by kronbichler** [comment]
And even if `SphericalManifold` would still be using the 3D, it would be correct.
